internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -1092,3 +1198,63 @@ func (r *BMACReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		Watches(&source.Kind{Type: &hivev1.ClusterDeployment{}}, handler.EnqueueRequestsFromMapFunc(mapClusterDeploymentToBMH)).
 		Complete(r)
 }
+
+func (r *BMACReconciler) formatMCSCertificateIgnition(mcsCert string) (string, error) {
+	var ignitionParams = map[string]string{
+		"SOURCE": fmt.Sprintf("data:text/plain;charset=utf-8;base64,%s", mcsCert),
+	}
+	tmpl, err := template.New("nodeIgnition").Parse(certificateAuthoritiesIgnitionOverride)
+	if err != nil {
+		return "", err
+	}
+	buf := &bytes.Buffer{}
+	if err = tmpl.Execute(buf, ignitionParams); err != nil {
+		return "", err
+	}
+	return buf.String(), nil
+}
+
+func (r *BMACReconciler) validateWorkerForDay2(log logrus.FieldLogger, agent *aiv1beta1.Agent) bool {
+	// Only worker role is supported for day2 operation
+	if agent.Spec.Role != models.HostRoleWorker || agent.Spec.ClusterDeploymentName == nil {
+		log.Debugf("Skipping spoke BareMetalHost reconcile for  agent %s/%s, role %s and clusterDeployment %s.", agent.Namespace, agent.Name, agent.Spec.Role, agent.Spec.ClusterDeploymentName)
+		return false
+	}
+	return true
+}
+
+func (r *BMACReconciler) getClusterDeploymentAndCheckIfInstalled(ctx context.Context, log logrus.FieldLogger, agent *aiv1beta1.Agent) (error, bool, *hivev1.ClusterDeployment) {
+	clusterDeployment := &hivev1.ClusterDeployment{}
+
+	cdKey := types.NamespacedName{
+		Namespace: agent.Spec.ClusterDeploymentName.Namespace,
+		Name:      agent.Spec.ClusterDeploymentName.Name,
+	}
+	var err error
+	if err = r.Get(ctx, cdKey, clusterDeployment); err != nil {
+		log.WithError(err).Errorf("failed to get clusterDeployment resource %s/%s", cdKey.Namespace, cdKey.Name)
+		return err, false, clusterDeployment
+	}
+
+	// If the cluster is not installed yet, we can't get kubeconfig for the cluster yet.
+	if !clusterDeployment.Spec.Installed {
+		log.Infof("ClusterDeployment %s/%s for agent %s/%s is not installed yet", clusterDeployment.Namespace, clusterDeployment.Name, agent.Namespace, agent.Name)
+		// If cluster is not Installed, wait until a reconcile is trigged by a ClusterDeployment watch event instead
+		return err, false, clusterDeployment
+	}
+	return err, true, clusterDeployment
+}
+
+func (r *BMACReconciler) getSecret(ctx context.Context, log logrus.FieldLogger, agent *aiv1beta1.Agent, clusterDeployment *hivev1.ClusterDeployment) (error, bool, *corev1.Secret) {
+	secret := &corev1.Secret{}
+	name := fmt.Sprintf(adminKubeConfigStringTemplate, clusterDeployment.Name)
+
+	err := r.Get(ctx, types.NamespacedName{Namespace: agent.Spec.ClusterDeploymentName.Namespace, Name: name}, secret)
+	if err != nil && k8serrors.IsNotFound(err) {
+		log.WithError(err).Errorf("failed to get secret resource %s/%s", agent.Spec.ClusterDeploymentName.Namespace, name)
+		// TODO: If secret is not found, wait until a reconcile is trigged by a watch event instead
+	} else if err != nil {
+		return err, true, secret
+	}
+	return err, false, secret
+}
```

Comment:
Instead of moving everything inside one big common function, I think it is better to have smaller functions dedicated to validate and return one particular type.